channels/offline-channel-js/src/OfflineBatchHandler.ts

```diff
@@ -0,0 +1,387 @@
+                                                        } else {
+                                                            let isRetriable = _isRetriable(evt, status);
+                                                            if (isRetriable) {
+                                                                result.state = eBatchSendStatus.Retry;
+                                                                _unloadProvider && _unloadProvider.addEvent(evt.id, evt, _itemCtx);
+
+                                                            } else {
+                                                                result.state = eBatchSendStatus.Drop;
+                                                            }
+                                                        }
+                                                        storeResolve(result);
+                                                        
+                                                    };
+                                                    return doAwaitResponse(sender(evt, onCompleteCallback, sync) as any, (res) => {
+                                                        // do nothing, because it should be resolved in callback
+                                                    });
+                                                }
+                                                // call resolve
+                                                result.state = eBatchSendStatus.Drop;
+                                                storeResolve(result);
+                                                
+                                            } catch (e) {
+                                                reject(e)
+                                            }
+                                        }
+                                        // this case should never happen because this function is called after we confirm that evt is available
+                                        // add resolve here in case
+                                        result.state = eBatchSendStatus.Complete;
+                                        storeResolve(result);
+
+                                    }
+                                    // if can not delete first, resolve
+                                    result.state = eBatchSendStatus.Failure;
+                                    result.data = res.reason;
+                                    storeResolve(result)
+                                
+                                });
+                            }
+                            // don't have any evts returned by getNext, resolve
+                            storeResolve(result);
+                            
+                        } catch(e) {
+                            reject(e);
+                        }
+                    });
+                    function storeResolve(result) {
+                        cb && cb(result);
```

Comment:
Just some edge case issue, we should wrap any callback in a try / catch, so that we still call (at least) the `resolve` "just in case" any callback fails for some uexpected reason.